File: Python/210__Course_Schedule_II.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        def helper(i):
            visited[i] = 1
            if i in graph:
                for j in graph[i]:
                    if visited[j] == 0:
                        if not helper(j):
                            return False
                    elif visited[j] == 1:
                        return False
            visited[i] = 2
            stack.insert(0, i)
            return True

        visited = [0] * numCourses
        graph = {}
        stack = []
        for pair in prerequisites:
            if pair[1] in graph:
                graph[pair[1]].add(pair[0])
            else:
                graph[pair[1]] = set([pair[0]])

        for i in range(numCourses):
            if visited[i] == 0:
                if not helper(i):
                    return []
        return stack

File: Java/210__Course_Schedule_II.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        boolean is_possible = true;
        Map<Integer, List<Integer>> adj_list = new HashMap<>();
        int[] degree = new int[numCourses];
        int[] result = new int[numCourses];
        for (int i = 0; i < prerequisites.length; i++) {
            int dest = prerequisites[i][0], src = prerequisites[i][1];
            List<Integer> lst = adj_list.getOrDefault(src, new ArrayList<Integer>());
            lst.add(dest);
            adj_list.put(src, lst);
            degree[dest]++;
        }
        Queue<Integer> queue = new LinkedList<Integer>();
        for (int i = 0; i < numCourses; i++)
            if (degree[i] == 0)
                queue.add(i);
        int i = 0;
        while (!queue.isEmpty()) {
            int node = queue.remove();
            result[i++] = node;
            if (adj_list.containsKey(node))
                for (Integer neighbor : adj_list.get(node)) {
                    degree[neighbor]--;
                    if (degree[neighbor] == 0)
                        queue.add(neighbor);
                }
        }
        if (i == numCourses)
            return result;
        return new int[0];

    }
}
